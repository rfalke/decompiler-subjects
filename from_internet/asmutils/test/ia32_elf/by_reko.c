// subject_seg08048000.c
// Generated by decompiling subject.exe
// using Reko decompiler version VERSION

#include "subject.h"

// 0804805C: Register int32 fn0804805C(Register Eq_3 esi)
// Called from:
//      fn08048082
int32 fn0804805C(Eq_3 esi)
{
	int32 edx_21 = 0x00;
	do
	{
		ci8 al_11 = esi->b0000;
		esi = (Eq_3) (esi + 1);
	} while (al_11 <= 33);
	while (al_11 > 0x2F && al_11 <= 0x3A)
	{
		edx_21 += (uint32) (al_11 - 0x30);
		al_11 = esi->b0000;
		esi = (Eq_3) (esi + 1);
		if (al_11 <= 0x2F || al_11 > 0x3A)
			return edx_21;
		edx_21 *= 0x0A;
	}
	return edx_21;
}

// 08048082: void fn08048082(Stack word32 dwArg00, Stack Eq_3 dwArg08, Stack Eq_3 dwArg0C, Stack Eq_3 dwArg10)
void fn08048082(word32 dwArg00, Eq_3 dwArg08, Eq_3 dwArg0C, Eq_3 dwArg10)
{
	ptr32 fp;
	struct Eq_38 * esp_179 = fp + 4;
	if (dwArg00 == 0x01)
	{
l08048264:
		int32 * esp_293 = esp_179 - 4;
		*esp_293 = 0x01;
		int32 ebx_295 = *esp_293;
		*esp_293 = 0x01;
		sys_exit(ebx_295);
	}
	word32 ebp_158 = 0x00;
	ptr32 esp_125 = fp + 0x0C;
	if (dwArg08->b0000 == 0x2D)
	{
		Eq_72 bl_16 = dwArg08[1];
		esp_179 = fp + 16;
		if (dwArg0C == null)
			goto l08048264;
		if (bl_16 != 122)
		{
			if (bl_16 != 110)
			{
				if (bl_16 != 116)
				{
					esp_179 = fp + 16;
					if (sys_lstat(dwArg0C, &g_t804829B) >= 0x00)
					{
						Eq_72 bl_77 = dwArg08[1];
						if (bl_77 != 115)
						{
							int32 ecx_84;
							if (bl_77 != 114)
							{
								if (bl_77 != 0x77)
								{
									byte * esi_102 = g_a804826D;
									do
									{
										Eq_72 al_105 = *esi_102;
										ui32 * esi_106 = esi_102 + 1;
										if (al_105 == bl_77)
										{
											ui32 eax_110 = *esi_106;
											if (eax_110 != (g_dw80482A3 & eax_110))
												goto l08048264;
											goto l0804825D;
										}
										esi_102 = (byte *) (esi_106 + 1);
									} while (al_105 != 0x00);
l0804804C:
									sys_exit(0x02);
								}
								ecx_84 = 0x01;
							}
							else
								ecx_84 = 0x00;
							esp_179 = fp + 16;
							if (sys_open(dwArg0C, ecx_84, dwArg0C) >= 0x00)
								goto l0804825D;
						}
						else if (g_dw80482AF == 0x00)
							goto l0804825D;
					}
				}
				else
				{
					esp_179 = fp + 16;
					if (sys_ioctl(fn0804805C(dwArg0C), 0x5401, 134513371) >= 0x00)
						goto l0804825D;
				}
				goto l08048264;
			}
			ebp_158 = 0x01;
		}
		if (ebp_158 == 0x00)
			goto l0804825D;
		goto l08048264;
	}
	if (dwArg08->b0000 != 33)
	{
		esp_125 = fp + 16;
		if (dwArg0C == null)
			goto l0804804C;
		if (dwArg0C->b0000 == 0x3D)
		{
l0804823E:
			memcmp(dwArg10, dwArg08, strlen(dwArg08) - ~0x00);
			int32 firstDifference_181 = __find_first_difference__(dwArg10, dwArg08);
			word32 eax_162 = 0x00;
			esp_179 = fp + 20;
			struct Eq_140 * esi_183 = dwArg10 + firstDifference_181;
			if (*((char *) (dwArg08 + firstDifference_181) - 1) == 0x00 && esi_183->bFFFFFFFF == 0x00)
				eax_162 = 0x01;
			if ((eax_162 ^ ebp_158) != 0x00)
				goto l0804825D;
			goto l08048264;
		}
		if (dwArg0C->b0000 == 0x3D21)
		{
l0804823B:
			ebp_158 = 0x01;
			goto l0804823E;
		}
		if (dwArg0C->b0000 == 0x2D)
		{
			Eq_72 ax_149 = dwArg0C[1];
			if (dwArg0C[3] != 0x00)
				goto l0804804C;
			if (ax_149 != 29029)
			{
				if (ax_149 != 0x656E)
				{
					if (ax_149 != 29799)
					{
						if (ax_149 != 25959)
						{
							if (ax_149 != 0x746C)
							{
								if (ax_149 != 0x656C)
									goto l0804804C;
								ebp_158 = 0x02;
							}
							Eq_174 eax_252 = fn0804805C(dwArg10);
							Eq_174 eax_263 = fn0804805C(dwArg08);
							esp_179 = fp + 20;
							if (ebp_158 != 0x02)
							{
								if (eax_263 >= eax_252)
									goto l08048264;
							}
							else if (eax_263 > eax_252)
								goto l08048264;
l0804825D:
							esp_179->dwFFFFFFFC = 0x01;
							sys_exit(0x00);
						}
						ebp_158 = 0x02;
					}
					Eq_174 eax_212 = fn0804805C(dwArg10);
					Eq_174 eax_223 = fn0804805C(dwArg08);
					esp_179 = fp + 20;
					if (ebp_158 != 0x02)
					{
						if (eax_223 > eax_212)
							goto l0804825D;
					}
					else if (eax_223 >= eax_212)
						goto l0804825D;
					goto l08048264;
				}
				goto l0804823B;
			}
			goto l0804823E;
		}
	}
	int32 * esp_140 = esp_125 - 4;
	*esp_140 = 0x03;
	int32 ebx_142 = *esp_140;
	*esp_140 = 0x01;
	sys_exit(ebx_142);
}

Eq_307 g_a804826D[] = // 0804826D
	{
	};
Eq_253 g_t804829B = // 0804829B
	{
	};
ui32 g_dw80482A3 = 0x00; // 080482A3
word32 g_dw80482AF = 0x00; // 080482AF
