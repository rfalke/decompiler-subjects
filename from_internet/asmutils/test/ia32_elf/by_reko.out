=== Using options:--aggressive-branch-removal
0804826D: warning: Expected sizes of arrays to have been determined by now
0804826D: warning: Expected sizes of arrays to have been determined by now
